include/tvm/relay/pass_manager.h

```diff
@@ -0,0 +1,393 @@
+// node with the same type. The Relay module is captured for optimizations as
+// most of the current Relay optimizations are module to module. Currying
+// sketches the optimization, i.e. how we want to mutate an AST, and it is
+// passed as packed functions that will be invoked when called by `run`.
+//
+// For example, PassFunc<Function> indicates we perform a Function to Function
+// transformation on the given Module.
+template <typename NodeT,
+          typename = std::enable_if<(std::is_same<NodeT, Module>::value ||
+                                     std::is_same<NodeT, Function>::value)>>
+using PassFunc =
+    runtime::TypedPackedFunc<runtime::TypedPackedFunc<NodeT(NodeT)>(
+        const Module& mod)>;
+
+class Pass;
+
+/*!
+ * \brief PassNode is the base type of differnt types of optimization passes.
+ * It is implemented by different pass subclasses at different granularity of
+ * Relay nodes.
+ */
+class PassNode : public RelayNode {
+ public:
+  /*! \brief The name of an optimization/analysis pass. */
+  std::string name;
+  /*! \brief The minimal optimization level that this pass will be enabled. */
+  int opt_level;
+
+  /*! \brief Set the context information for a pass. */
+  void SetContext(const PassContext& pass_ctx) {
+    pass_ctx_ = pass_ctx;
+  }
+
+  /*!
+   * \brief Get the required passes for this pass as a vector of std::string.
+   */
+  virtual std::vector<std::string> Required() const = 0;
+
+  /*!
+   * \brief Execute the optimization pass using a functor. This functor invokes
+   *        the `run` method to perform a real optimization on a certain type
+   *        of node.
+   *
+   * \param mod The module that an optimization pass runs on.
+   *
+   * \return The updated module.
+   */
+  virtual Module operator()(const Module& mod) const = 0;
+
+  /*!
+   * \brief Execute the optimization pass. This is function should be specilized
+   *        for different types of Relay nodes. For example, we mainly allow
+   *        transformation of from Module/Function to Module/Function. Note that
+   *        the module will be updated.
+   *
+   * \param mod The module that an optimization pass runs on.
+   *
+   * \return Return the updated module.
+   */
+  virtual Module Run(const Module& mod) const = 0;
+
+  void VisitAttrs(tvm::AttrVisitor* v) override {
+    v->Visit("name", &name);
+    v->Visit("opt_level", &opt_level);
+  }
+
+  static constexpr const char* _type_key = "relay.Pass";
+  TVM_DECLARE_BASE_NODE_INFO(PassNode, RelayNode);
+
+ protected:
+  /*!
+   * \brief The context information that is used to help perform a given pass.
+   */
+  PassContext pass_ctx_;
+};
+
+class Pass : public NodeRef {
+ public:
+  Pass() = default;
+  explicit Pass(NodePtr<tvm::Node> p) : NodeRef(p) {}
+
+  PassNode* operator->() const {
+    return static_cast<PassNode*>(this->node_.get());
+  }
+
+  using ContainerType = PassNode;
+};
+
+class ModulePass;
+
+/*!
```

Comment:
Module pass and FunctionPass are implementations, their signature should sit inside the cc file instead of header file.

Reply:
We should only leave the Constructor function CreateModulePass in the header